### PR TITLE
net: if: Rework interface locking 

### LINKED ISSUE #50875: net: ip: race in access to writable net_if attributes
**Describe the bug**

The net_if API may be concurrently accessed by several threads. Write access to certain net_if attributes (notably link-layer address) is therefore guarded by a mutex lock.
    
Locking has only been introduced for write access while non-atomic read access does not seem to be guarded. This seems to create a race condition during concurrent read and write access to net_if attributes unless I'm overlooking something that makes locking unnecessary in this special case...
    
As an aside: Currently all instances of net_if are guarded by a single global lock although there's no danger of race conditions across instances. It may make sense to introduce more granular per-instance locking of net_if rather than global locking.

**To Reproduce**

This bug has been found analytically. A sample patch will be provided that explains what a solution could look like.

**Expected behavior**

Memory that can be written and read concurrently by independent threads should be properly guarded against race conditions.

**Impact**

As the race condition is in code paths that may be executed with high frequency (e.g. net packet code), net packets might be corrupted when updating net_if attributes concurrently.

Example: The link layer address of an interface needs to be changed when switching from extended to short source addresses in the IEEE 802.15.4 protocol after associating to a PAN without taking the interface down for concurrent TX. Socket code will add the source address of a package on-the-fly by accessing the current link-layer address of the interface. Accessing the link-layer address is not safe, IMHO, while it is being written. So packets being created by a TX thread while the RX thread responds to a (dis)association response from a remote coordinator which switches the net_if's link-layer address from extended to short (or the opposite) may corrupt the outgoing packet's link-layer address when RX/TX run on different CPUs or RX get's preempted in the middle of updating the ll address.

**Logs and console output**

n/a

**Environment (please complete the following information):**

current mainline

**Additional context**

I'll attach a patch to this bug report which introduces a partial tentative fix by exposing the net_if mutex via internal API methods. These methods will then be used to exemplify my understanding of read-time locking in some frequently used functions of the net IP stack.
    
The patch is not intended as a fix. Many such fixes would have to be introduced to protect all critical sections with read access to writable parts of the net_if struct unless I'm overlooking something (which is entirely possible!).

Unfortunately this fix seems to introduce enough lag to **break several timing tests** in twister. So advice on whether and how this should be approached is highly appreciated.

### PR DESCRIPTION
**Pull request by rlubos**
Instead of having a single global mutex, covering, in most cases
unrelated, access to all network interfaces, introduce a separate mutex
for each network interface. Details in the commit messages.

Fixes #50875

### CODE REVIEW

**Review by jukkar — DISMISSED** [review]
Nice work :+1:

**Comment by carlescufi** [comment]
@fgrandel would you mind taking a look at this PR, since you opened the corresponding issue? thanks!

**Comment by fgrandel** [comment]
> @fgrandel would you mind taking a look at this PR, since you opened the corresponding issue? thanks!

@carlescufi Sure - will do.

**Review by ghost — DISMISSED** [review]
My questions should not block this PR. In any case this is an obvious improvement over the status quo.

This PR only partially fixes #50875, though. If the RX pool is exhausted and one thread needs TX on the same iface to unblock RX - or the other way round, a deadlock may still occur. See the detailed analysis given in https://github.com/zephyrproject-rtos/zephyr/pull/50905#issuecomment-1287864547. For this to be solved independent per-pool locks would have to be introduced.

File: include/zephyr/net/net_if.h
Comment (rlubos) [reply include/zephyr/net/net_if.h:530]: > What is the performance overhead of a mutex compared to a spinlock in these cases?

I don't know to be honest, for sure this is something that could be measured to see if there's a room for a possible enhancement.

> Are we sure that none of the protected sections may ever be called from an ISR context?

That's a good point. It actually seems, that changes in `net_pkt_set_iface()` have violated this requirement. There is a short code path, between calling `net_recv_data()` and queueing the packet, when `net_pkt_set_iface()` could be called from the IRQ context (this depends on the driver design). I need to rework this func to get rid of the locking.

> So maybe a combination of two mechanisms might be adequate?

To be honest, any combnation of two or more locking mechanisms sounds like a little nightmare to me. Combining the global mutex (protecting shared resources) with the local one in this PR gave me a bit of a headache already.

File: subsys/net/ip/net_if.c
Comment (jukkar) [reply subsys/net/ip/net_if.c:2052]: After looking the code, I am just wondering do we really need to have a lock in maddr_leave/join functions at all.

**Comment by rlubos** [comment]
Updated `net_if_ipv*_maddr_join()/leave()` APIs and removed lock from `net_pkt_set_iface()`.

File: include/zephyr/net/net_if.h
Comment (jukkar) [inline include/zephyr/net/net_if.h:629]: Is there a typo here?

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by fgrandel** [comment]
> The thing I don't follow, is that net_if code makes no pkt/buffer allocation so it's not clear to me where could the net_if code actually block because of this.

IIRC the situation in the test was roughly like this: The background networking thread caused a lot of RX resources to be blocked (queued) due to initial MLD, NS, ND, etc. But when the networking thread tried to allocate TX packages to handle the incoming packages it got starved as the test thread managed to block remaining TX resources in the meantime. Deadlock occurred because the test thread was waiting itself for some network event (assignment of an IP address or something - don't remember the details any more) which would then never happen.

Just pointing it out as I saw the network stack on real devices initially consume a lot of package resources exactly at the same time that userland processes gain control for the first time and may block TX resources in their own initialization routine while waiting for the network to become available.

Reserving a larger TX/RX thread will solve this but it's maybe not so obvious to users? Detecting the deadlock in the network thread (waiting for TX resources with timeout) and throwing a good error message may be a good enough solution then...

**Comment by rlubos** [comment]
> > The thing I don't follow, is that net_if code makes no pkt/buffer allocation so it's not clear to me where could the net_if code actually block because of this.
> 
> IIRC the situation in the test was roughly like this: The background networking thread caused a lot of RX resources to be blocked (queued) due to initial MLD, NS, ND, etc. But when the networking thread tried to allocate TX packages to handle the incoming packages it got starved while holding on to the mutex as the test thread managed to block remaining TX resources in the meantime. Deadlock occurred because the test thread was waiting itself for some network event or needed some network resource protected by the mutex (some networking function call, RX or assignment of an IP address or something - don't remember the details any more) which would then never happen.
> 
> Just pointing it out as I saw the network stack on real devices initially consume a lot of package resources exactly at the same time that userland processes gain control for the first time and may block TX resources in their own initialization routine while waiting for the network to become available.
> 
> Reserving a larger TX/RX thread will solve this but it's maybe not so obvious to users? Detecting the deadlock in the network thread (waiting for TX resources with timeout) and throwing a good error message may be a good enough solution then...

On second thought, I think that this might've happen because of the loopback interface nature - it does not send the packet to the wire, but creates a **copy** of it (so it's an exception in terms of packet allocation) and sends it back to the stack. With `iface` buffer being held in `net_pkt_set_iface()` on the RX path (which please note is no longer the case after recent change), we may hit the RX packet shortage.

I think though that "deadlock" is a too strong description for the problem - yes, it may seems so as the whole boot sequence blocked for the 30 seconds, but this was due to DAD procedure failure. The actual impact of the mutex being held was packet allocation failure (there's a 100 ms timeout in the driver) so the packet did not reach back to the stack.

**Review by carlescufi — APPROVED** [review]
Reporter @fgrandel previously approved this PR before a rebase, so merging.

**Pull request merged**
